external-crates/move/move-compiler/src/typing/translate.rs

```diff
@@ -59,16 +59,45 @@ fn modules(
     context: &mut Context,
     modules: UniqueMap<ModuleIdent, N::ModuleDefinition>,
 ) -> UniqueMap<ModuleIdent, T::ModuleDefinition> {
-    modules.map(|ident, mdef| module(context, ident, mdef))
+    let mut all_new_friends = BTreeMap::new();
+    let mut typed_modules = modules.map(|ident, mdef| {
+        let (typed_mdef, new_friends) = module(context, ident, mdef);
+        for (pub_package_module, loc) in new_friends {
+            all_new_friends
+                .entry(pub_package_module)
+                .or_insert_with(BTreeMap::new)
+                .insert(
+                    ident,
+                    Friend {
+                        attributes: UniqueMap::new(),
+                        loc,
+                    },
+                );
+        }
+        typed_mdef
+    });
+
+    for (mident, friends) in all_new_friends {
+        let mdef = typed_modules.get_mut(&mident).unwrap();
+        // point of interest: if we have any new friends, we know there can't be any
+        // "current" friends becahse all thew new friends are generated off of
+        // `public(package)` usage, which disallows other friends.
+        mdef.friends = friends
+            .into_iter()
+            .collect::<UniqueMap<ModuleIdent, Friend>>();
```

Comment:
Won't this cause a crash if the module that calls the public package function is a friend?

Reply:
Let's add a test to be sure
```
module 0x42::m {
  friend 0x42::other;
  public(package) fun foo() {}
}
module 0x42::other {
  fun call_foo() { 0x42::m::foo() }
}
```

Reply:
comment in the code explains:

```
        // point of interest: if we have any new friends, we know there can't be any
        // "current" friends becahse all thew new friends are generated off of
        // `public(package)` usage, which disallows other friends.
 ```
 
 We can clobber this because of the disjoint usage requirement.